Cria novo Evento rápido

anacunha opened the pull request:
### Necessidades

- [X] Um título
- [X] O momento de começo
- [X] O momento de fim
- [X] Lista de convidados por email
- [X] Uma descrição
- [x] A agenda que aquele evento possui
- [ ] O dono do evento é o usuário logado naquele instante

### Restrições

- [X] título não pode ser nulo, nem vazio e tem no máximo 100 caracteres
- [X] O momento de começo precisa ser no futuro
- [X] O momento de fim precisa ser no futuro e depois do começo
- [x] Lista de convidados por email só pode ter email de usuário que está cadastrado
- [X] Descrição é obrigatória e tem no máximo 255 caracteres
- [x] A agenda associada ao evento é obrigatória

### Resultado Esperado

Um novo evento criado associado ao usuário logado

anacunha committed:
Controller e DTO para criação de novo Evento rápido

anacunha committed:
Salva novo Evento rápido

anacunha commented on src/main/java/br/com/deveficiente/calendario/controller/form/NovoEventoForm.java:
```diff
@@ -0,0 +1,112 @@
+
+        final List<Usuario> convidados = emailsConvidados.stream()
+                .map(email -> usuarioRepository.findByLogin(email).orElseThrow(usuarioNotFoundException(email)))
+                .collect(Collectors.toList());
```
Aqui fiquei um pouco na dúvida se não seria melhor retornar um erro de validação do que lançar uma exceção 🤔

asouza commented on src/main/java/br/com/deveficiente/calendario/controller/EventoController.java:
```diff
@@ -0,0 +1,45 @@
+    @Transactional
+    public void cria(@RequestBody @Valid final NovoEventoForm form) {
+        // TODO Receber Usuário pelo método após habilitar autenticação via token
+        final Usuario usuario = usuarioRepository.findByLogin("ana@email.com").get();
```
mandou bem aqui :). Receber o usuário logado via token vira só um detalhe e não interfere em nada na lógica daí para baixo. Depois você altera e tudo continua funcionando.

asouza commented on src/main/java/br/com/deveficiente/calendario/controller/EventoController.java:
```diff
@@ -0,0 +1,45 @@
+
+    @PostMapping("/api/evento")
+    @Transactional
+    public void cria(@RequestBody @Valid final NovoEventoForm form) {
```
final no argumento foi um toque chique.

asouza commented on src/main/java/br/com/deveficiente/calendario/controller/form/NovoEventoForm.java:
```diff
@@ -0,0 +1,112 @@
+
+    public Evento toModel(final AgendaRepository agendaRepository, final UsuarioRepository usuarioRepository,
+                          final Usuario dono) {
+        final Agenda agenda = agendaRepository.findById(idAgenda).orElseThrow(agendaNotFoundException(idAgenda));
```
Exception ou erro de validação depende do requisito, realmente não estava especificado. No final, acho que ficou legal. Esperamos sempre uma agenda que existe mesmo. O único detalhe que eu faria diferente é: um id errado, do meu ponto de vista, pode ser interpretado como um argumento inválido. Então eu usaria o que já existe e lançaria uma IllegalArgumentException. 

Por conta disso aqui => *Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*

anacunha replied:
Achei um bom argumento, depois a gente fica com sistema com inúmeras exceções que ninguém conhece ao certo (muito menos nosso cliente). Valeu!

asouza commented on src/main/java/br/com/deveficiente/calendario/controller/form/NovoEventoForm.java:
```diff
@@ -0,0 +1,112 @@
+        final Agenda agenda = agendaRepository.findById(idAgenda).orElseThrow(agendaNotFoundException(idAgenda));
+
+        final List<Usuario> convidados = emailsConvidados.stream()
+                .map(email -> usuarioRepository.findByLogin(email).orElseThrow(usuarioNotFoundException(email)))
```
Já aqui merece um validador :). Pode ser que o código cliente passe um email de usuário que não existe e aí é necessário retornar o erro de validação.

anacunha replied:
Feito! E acho que faz mais sentido, a pessoa pode não saber que alguém que ela convidou não está cadastrado.

Só não amei o fato de repetir o código que faz o `findByLogin` pra todos emails dos convidados no validador e na hora de criar o Evento.

asouza commented on src/main/java/br/com/deveficiente/calendario/model/Evento.java:
```diff
@@ -0,0 +1,63 @@
+    private final String descricao;
+
+    @ManyToMany
+    private final List<Usuario> convidados;
```
Aqui, em vez de usar um List, sugiro usar um Set :). Já que não temos convidados repetidos no mesmo evento. Pode considerar o email para critério de comparação do usuário. Se é que você já não fez isso :).

anacunha replied:
Boa! Não tinha pensado nisso, lembro que fizemos a mesma coisa pra a lista de convidados do bolão no workshop. Deixei o `equals` e `hashCode` usando apenas o e-email, conforme sugerido.

asouza replied:
É... confesso que esse é um efeito colateral sim :(. A solução alternativa seria manter o código de validação dos emails mais perto do código que executa  a lógica em si. Só que aí você ia ter que tratar a falha de validação meio que na mão... Nessa troca, eu prefiro a query duplicada :). Caso isso vire um problema, da para pensar em algumas coisas:

* Query que busca por todos os usuarios com emails ```in(listaEmails)```. Se a collection retornada tiver tamanho diferente da coleção do form, quer dizer que alguém não existe. 
* Colocar algum cache bobo em memória ou mais parrudo com um redis da vida.

asouza commented on src/main/java/br/com/deveficiente/calendario/model/Evento.java:
```diff
@@ -0,0 +1,63 @@
+
+    public Evento(@NotBlank @Length(max = 100) final String titulo, @Future @NotNull final LocalDateTime inico,
+                  @Future @NotNull final LocalDateTime termino, @NotBlank @Length(max = 255) final String descricao,
+                  final List<Usuario> convidados, @NotNull final Agenda agenda,
```
Aqui me surgiu uma dúvida que eu tinha deixado passar no formulário :). No cartão eu só disse que os emails precisavam ser de usuários que existem. Por outro lado não falei que precisava ter pelo menos um convidado. 

No construtor você obriga a passagem dos convidados, mas no formulário não obriga ter pelo menos um email. Já que você obrigou aqui(e tá tudo bem), vale colocar um @Size lá na lista de emails do form e daqui da entidade.

anacunha replied:
A minha ideia era permitir que a lista de convidados fosse passada vazia ou nula no form, não acho que faça sentido forçar a ter convidados no evento.

Inicializei a lista no form e o set no evento automaticamente pra tentar resolver isso. Testei aqui com lista vazia e nula (omitida, na verdade) e parece que deu certo.

O único ponto que pode dar ruim é se a lista no `NovoEventoForm` for explicitamente setada como nula, dai quebra o construtor do `Evento` que agora tá usando `addAll`. Ao não passar a lista chamando o método POST, parece que o setter dela não foi chamado, mas ainda assim existiria essa possibilidade.

Como alternativa eu vejo:

- Não usar um construtor para criar o `Evento`
- Verificar se a lista passada para o construtor é nula antes de chamar o `addAll`

Confesso não que não gosto muito de nenhuma das duas. Talvez o melhor seja pelo menos pedir que a lista não seja nula no form. O que você acha?

anacunha replied:
O validador da lista de e-mails tb quebra se ela for explicitamente nula 😞 

Me parece que faz sentido deixar ela como `@NotNull` mesmo

asouza commented on src/main/java/br/com/deveficiente/calendario/validator/PeriodoEventoValidator.java:
```diff
@@ -0,0 +1,27 @@
+        final LocalDateTime inico = novoEventoForm.getInico();
+        final LocalDateTime termino = novoEventoForm.getTermino();
+
+        if (!termino.isAfter(inico)) {
```
Esse código está bem de boa. Por outro lado, inicio e termino são atributos do form. Em vez fazer a lógica em cima do retorno dos dados do mesmo objeto, você pode criar um método que já te retorne o boolean em relação ao término depois do começo. Em geral, maximizar o uso do **this** leva a deixar o estado perto do comportamento e com isso o código fica mais coeso. E coesão tende a facilitar o entendimento :).

anacunha replied:
Boa! Criei um método `isTerminoAfterInicio()` no form, mas precisei usar `@JsonAutoDetect(isGetterVisibility = JsonAutoDetect.Visibility.NONE)` pra ele não aparecer no Swagger. É isso mesmo?

asouza replied:
Ele pega por conta do padrão do nome, que começa com is. Da para mandar um @JsonIgnore em cima do método ou trocar o nome :P. Nessas classes de DTO ta tudo bem ser um pouco mais flexível no design.

asouza reviewed (changes requested):
Aqui tem apenas alguns detalhes que acho que vale a reflexão.

anacunha committed:
Mudanças ao salvar Evento rápido

- Validador da lista de emails dos convidados
- Evento usa um conjunto para guardar lista de convidados
- Lança IllegalArgumentException se id da agenda não existe

asouza commented on src/main/java/br/com/deveficiente/calendario/controller/form/NovoEventoForm.java:
```diff
@@ -38,7 +37,7 @@
     private String descricao;
 
-    private List<@Email String> emailsConvidados;
+    private List<@Email String> emailsConvidados = new ArrayList<>();
```
isso aí Não é que precisa de @NotNull, é que sempre que você puder iniciar a variável com um valor é melhor(claro que nem sempre é possível). No caso de coleções, é só jogar a coleção vazia :).

asouza reviewed (approved):
Boa! Deixei só um comentário relativo ao que poderia ser feito na validação do email :).
